File: js/hello_world.js
// A simple example of using Machnet that sends the message "Hello World!" over
// the network.
//
// Requirements: npm install ref-napi ffi-napi ref-struct-napi
//
// Usage: Assuming we have two servers (A and B), where Machnet is running on both
// IP 10.0.255.100 at server A, and IP 10.0.255.101 at server B.
//
// On server A: node hello_world.js --local_ip 10.0.255.100
// On server B: node hello_world.js --local_ip 10.0.255.101 --remote_ip 10.0.255.100
//
// If everything works, server A should print "Received message: Hello World!"

const ref = require('ref-napi');
const commander = require('commander');
const chalk = require('chalk');
const {machnet_shim, MachnetFlow_t} = require('./machnet_shim');

function customCheck(condition, message) {
  if (!condition) {
    console.log(chalk.red('Error: ' + message));
    process.exit(1);
  } else {
    console.log(chalk.green('Success: ' + message));
  }
}

const kHelloWorldPort = 888;
commander.option('-l, --local_ip <ip>', 'Local IP address')
    .option('-r, --remote_ip <ip>', 'Remote IP address')
    .parse(process.argv);

const options = commander.opts();
if (!options.local_ip) {
  console.log(chalk.red('Error: local_ip is required'));
  process.exit(1);
}
console.log(options);

// Main logic
var ret = machnet_shim.machnet_init();
customCheck(ret === 0, 'machnet_init()');

var channel_ctx = machnet_shim.machnet_attach();
customCheck(channel_ctx !== null, 'machnet_attach()');

if (options.remote_ip) {
  // Client
  const flow = new MachnetFlow_t();
  ret = machnet_shim.machnet_connect(
      channel_ctx, ref.allocCString(options.local_ip),
      ref.allocCString(options.remote_ip), kHelloWorldPort, flow.ref());
  customCheck(ret === 0, 'machnet_connect()');

  const msg = 'Hello World!';
  const msg_buffer = Buffer.from(msg, 'utf8');
  ret = machnet_shim.machnet_send(channel_ctx, flow, msg_buffer, msg_buffer.length);
  if (ret === -1) {
    console.log(chalk.red('Error: machnet_send() failed'));
  } else {
    console.log(chalk.green('Message sent successfully'));
  }
} else {
  // Server
  console.log('Waiting for message from client');
  ret = machnet_shim.machnet_listen(
      channel_ctx, ref.allocCString(options.local_ip), kHelloWorldPort);
  customCheck(ret === 0, 'machnet_listen()');

  function receive_message() {
    const buf = Buffer.alloc(1024);
    const flow = new MachnetFlow_t();
    const bytesRead =
        machnet_shim.machnet_recv(channel_ctx, buf, buf.length, flow.ref());

    if (bytesRead === -1) {
      console.log(chalk.red('Error: machnet_recv() failed'));
    } else if (bytesRead === 0) {
      setTimeout(receive_message, 10);
    } else {
      const receivedMsg = buf.toString('utf8', 0, bytesRead);
      console.log(`Received message: ${receivedMsg}`);
      receive_message();
    }
  }

  receive_message();
}

File: examples/hello_world.cc
/**
 * @file main.cc
 * Simple hello world application using only Machnet public APIs
 * Usage:
 *  - First start the server: ./hello_world --local=<local IP>
 *  - Client:
 *    ./hello_world --local=<local IP> --remote=<server IP> --is_client=1
 */

#include <gflags/gflags.h>
#include <machnet.h>

#include <array>

DEFINE_string(local, "", "Local IP address");
DEFINE_string(remote, "", "Remote IP address");

static constexpr uint16_t kPort = 31580;

// assert with message
void assert_with_msg(bool cond, const char *msg) {
  if (!cond) {
    printf("%s\n", msg);
    exit(-1);
  }
}

int main(int argc, char *argv[]) {
  google::ParseCommandLineFlags(&argc, &argv, true);

  int ret = machnet_init();
  assert_with_msg(ret == 0, "machnet_init() failed");

  void *channel = machnet_attach();
  assert_with_msg(channel != nullptr, "machnet_attach() failed");

  ret = machnet_listen(channel, FLAGS_local.c_str(), kPort);
  assert_with_msg(ret == 0, "machnet_listen() failed");

  printf("Listening on %s:%d\n", FLAGS_local.c_str(), kPort);

  if (FLAGS_remote != "") {
    printf("Sending message to %s:%d\n", FLAGS_remote.c_str(), kPort);
    MachnetFlow flow;
    std::string msg = "Hello World!";
    ret = machnet_connect(channel, FLAGS_local.c_str(), FLAGS_remote.c_str(),
                          kPort, &flow);
    assert_with_msg(ret == 0, "machnet_connect() failed");

    const int ret = machnet_send(channel, flow, msg.data(), msg.size());
    if (ret == -1) printf("machnet_send() failed\n");
  } else {
    printf("Waiting for message from client\n");
    size_t count = 0;

    while (true) {
      std::array<char, 1024> buf;
      MachnetFlow flow;
      const ssize_t ret = machnet_recv(channel, buf.data(), buf.size(), &flow);
      assert_with_msg(ret >= 0, "machnet_recvmsg() failed");
      if (ret == 0) {
        usleep(10);
        continue;
      }

      std::string msg(buf.data(), ret);
      printf("Received message: %s, count = %zu\n", msg.c_str(), count++);
    }
  }

  return 0;
}
